src/permutation.rs

```diff
@@ -240,68 +341,196 @@ impl Permutation {
         }
     }
     fn apply_slice_fwd<T: Clone, D>(&self, vec: D) -> Vec<T>
-        where D: Deref<Target = [T]>
+    where
+        D: Deref<Target = [T]>,
     {
-        self.indices
-            .iter()
-            .map(|&idx| vec[idx].clone())
-            .collect()
+        self.indices.iter().map(|&idx| vec[idx].clone()).collect()
     }
 
     fn apply_slice_bkwd<T: Clone, D>(&self, vec: D) -> Vec<T>
-        where D: Deref<Target = [T]>
+    where
+        D: Deref<Target = [T]>,
     {
         let mut other: Vec<T> = vec.to_vec();
         for (i, idx) in self.indices.iter().enumerate() {
             other[*idx] = vec[i].clone();
         }
         return other;
     }
+
+    // For the in place methods, we apply each cycle in the permutation in turn, marking the indices with their MSB when
+    // they have been resolved. The MSB will always be unset as long as n <= isize::max_value().
+    // This way, we can recover the original indices in O(n) and perform no heap allocations.
+
+    #[inline(always)]
+    fn mark_idx(idx: usize) -> usize {
+        idx ^ isize::min_value() as usize
+    }
+
+    #[inline(always)]
+    fn idx_is_marked(idx: usize) -> bool {
+        (idx & (isize::min_value() as usize)) != 0
+    }
+
+    fn apply_slice_bkwd_in_place<T>(&mut self, slice: &mut [T]) {
+        for idx in self.indices.iter() {
```

Comment:
Maybe we should just assert about the size of the permutation being less than the limit, and only use a `debug_assert!()` for individually checking each index, so it will be optimized away in release builds.